scvid/module/_probabilisticpca.py

```diff
@@ -123,22 +126,30 @@ def model(self, x_ng: torch.Tensor):
                 )
 
     @pyro_method
-    def guide(self, x_ng: torch.Tensor):
+    def guide(self, x_ng: torch.Tensor) -> None:
         if self.ppca_flavor == "marginalized":
             return
 
         if self.transform is not None:
             x_ng = self.transform(x_ng)
 
         with pyro.plate("cells", size=self.n_cells, subsample_size=x_ng.shape[0]):
-            z_loc_nk = (x_ng - self.mean_g) @ self.L_gk
+            WX_kn = self.W_kg @ (x_ng - self.mean_g).T
+            z_loc_nk = torch.linalg.solve(self.M_kk, WX_kn).T
 
             if self.ppca_flavor == "diagonal_normal":
                 pyro.sample("z", dist.Normal(z_loc_nk, self.z_scale_k).to_event(1))
             elif self.ppca_flavor == "multivariate_normal":
+                M_inv_kk = torch.linalg.inv(self.M_kk)
+                # This function skips the (slow) error checking and error message construction
+                # of torch.linalg.cholesky(), instead directly returning the LAPACK error codes
+                # as part of a named tuple (L, info).
+                z_scale_tril_kk, _ = torch.linalg.cholesky_ex(
+                    self.sigma**2 * M_inv_kk
+                )
```

Comment:
Consider trying out the following stabilizing tricks in case you still run into NaNs:

(1) Try `self.sigma**2 * M_inv_kk + self.jitter * torch.eye(self.k_components, device=self.sigma.device)` as the Cholesky argument;

(2) Try `z_scale_tril_kk.view(-1)[::(self.k_components + 1)] += self.jitter` post Cholesky.

(3) Add the jitter directly to `W_kg @ W_kg.T` in `self.M_kk`.

Set `self.jitter` to 1e-6 initially, and potentially increase to 1e-3.